kaveh808/kons-9
# OpenGL mouse picking with ray casting

Branch [opengl-mouse-picking-with-ray-casting](https://github.com/kaveh808/kons-9/tree/opengl-mouse-picking-with-ray-casting) has some code to experiement with with OpenGL mouse picking using ray casting.

### How to test it out

1. Left click to select a single shape. In the case when multiple shapes lie in the path of the ray, then subsequent clicks select each shape in turn. This allows cycling amongst many closely overlapping objects. @kaveh this behaviour was arbitrarily done using blender as a reference - we can discuss and change it to suit to whatever works best for kons-9.

2. Left click while pressing the `shift` key to add an item to the selection.

3.  To write custom selection behaviour, see [test/demo-object-picking.lisp](https://github.com/kaveh808/kons-9/blob/opengl-mouse-picking-with-ray-casting/test/demo-object-picking.lisp)

### TODOs:

1. `triangles-list` in file `polyhedron.lisp` will only work for convex polyhedrons but we need to be able to intersect for all cases.
2. Currently, we triangulate each shape on each pick. Try to re-use triangulation for subsequent picks if a shape has not changed.
3. Only tried with polyhedrons which have a triangulation routine. There might be other shapes which need to be dealt with?


### How it tries to pick shapes

1. The first step in picking is to use `glu:un-project` to determine the **far** point under the mouse
2. Next, we extract the current **camera position** from the current modelview matrix using `(gl:get-double :modelview-matrix)`
3. The above two steps constitutes our `ray` from the camera to the far plane under the mouse
4. At this point, we create an instance of the `origin` library's `origin.geometry.ray` which gives us a **ray** object
5. Next we cycle through each objecet in the scene and do the following:
6. Get the axis aligned bounding box (*aabb*) of the shape in world coordinates
7. For the above **aabb**, we create an instance of the `origin` library's `origin.geometry.aabb`
8. Next we use the `origin` library function `origin.geometry:raycast-aabb` to intesect the above `ray` and `aabb` objects
9. If a shape's **aabb** did not intersect, we add it to a list called `xs-miss` and move on to the next shape.
10. If a shape's **aabb** does intersect, we further refine the intersection test by triangulating the shape and intersect the ray with each triangle for better certainty if the shape does intersect.
11. After intersecting each shape in the scene, we create a list called `xs-hit` containing a list of shapes sorted by distance from the camera. Anoter list `xs-miss` contains all shapes which did not intersect.
12. Finally we use `xs-hit`, `xs-hit` along with another list `xs-current` to decide which shapes should be selected in the scene. See [test/demo-object-picking.lisp](https://github.com/kaveh808/kons-9/blob/opengl-mouse-picking-with-ray-casting/test/demo-object-picking.lisp) for more

Will look forward to comments and suggestions.

(Below is older stuff, not relevant anymore)

~~This demo picks objects whenever the mouse hovers over them which is done as follows:~~

~~1. Use a flag (`*do-mouse-pick*`) to detect mouse move~~
~~2. Immediately after each **draw scene** the object picker described above kicks in if the flag `*do-mouse-pick*` was set~~

~~### Picking resolution/accuracy~~

~~Since the ray is intersected with the **aabb** (axis aligned bounding box) an objects gets picked whenever it intersects the bounding box of a shape. As you will observe, this is not very accurate.~~

~~### Next steps~~

~~This was more like an fun experiment.  If this type of picking might be of use to kons-9 (at-least until vulkan comes in), we can:~~

~~1. Improve the picking accuracy: If an object's **aabb** does get picked, attempt further intersections of the ray with the actual shape and not just the aabb~~
~~2. Design an api to enable/disable picking with useful options such as on mouse clicks / double clicks / hover, etc...~~

~~### Notes~~

~~1. For this demo, all objects that are in the path of the ray get picked, not only the closest one. This behaviour is easy to change and pick only the object closest to the camera. This can be done using the number returned by intersect funtion (`origin.geometry:raycast-aabb`) which indicates the distance from the camera.~~

~~Since this was an experiment, this is just a draft pull request to view code diffs and get feedback (rather than a normal pull request)~~
